test/helpers/sockets.rb

```diff
@@ -0,0 +1,435 @@
+# frozen_string_literal: true
+
+require 'socket'
+require 'timeout'
+
+module TestPuma
+
+  READ_TIMEOUT = 10
+
+  module SktPrepend
+    def get_body(path = nil, dly: nil, len: nil, timeout: 10)
+      fast_write get_req(path, dly: dly, len: len)
+      read_body timeout
+    end
+
+    def get_response(path = nil, dly: nil, len: nil, timeout: 10)
+      fast_write get_req(path, dly: dly, len: len)
+      read_response timeout
+    end
+
+    def read_body(timeout = 10)
+      response = read_response timeout
+      response.split("\r\n\r\n").last
+    end
+
+    def read_response(timeout = 10)
+      content_length = nil
+      response = ''.dup
+      t_st = Time.now
+      loop do
+        begin
+          chunk = read_nonblock(16_384, exception: true)
+          if chunk.is_a? String
+            response << chunk
+            if (content_length ||= (t = response[/Content-Length: (\d+)/i , 1]) ? t.to_i : nil)
+              return response if response.split("\r\n\r\n", 2).last.bytesize == content_length
+            else
+              return response if response.split("\r\n\r\n", 2).length == 2
+            end
+          end
+        rescue EOFError # for debugging, intermittent
+          raise EOFError
+        rescue ::IO::WaitReadable
+        ensure
+          sleep 0.0002
+          if timeout < Time.now - t_st
+            raise ::Timeout::Error, 'Client Read Timeout'
+          end
+        end
+      end
+    end
+
+    def get(path = nil, dly: nil, len: nil)
+      fast_write get_req(path, dly: dly, len: len)
+    end
+
+    def write(str)
+      fast_write str
+    end
+
+    def <<(str)
+      fast_write str
+    end
+
+    def fast_write(str)
+      n = 0
+
+      while true
+        begin
+          n = syswrite str
+        rescue Errno::EAGAIN, Errno::EWOULDBLOCK => e
+          raise e unless IO.select(nil, [io], nil, 5)
+          retry
+        rescue Errno::EPIPE, SystemCallError, IOError => e
+          raise e
+        end
+
+        return if n == str.bytesize
+        str = str.byteslice(n..-1)
+      end
+    end
+
+    private
+
+    def get_req(path = nil, dly: nil, len: nil)
+      req = "GET /#{path} HTTP/1.1\r\n".dup
+      req << "Dly: #{dly}\r\n" if dly
+      req << "Len: #{len}\r\n" if len
+      req << "\r\n"
+    end
+  end
+
+  class SktTCP < ::TCPSocket
```

Comment:
Hmm... it would definitely be confusing if we just called it `TCPSocket`...

Maybe `ExtendedTCPSocket`, `ExtendedUnixSocket` et cetera